### ISSUE TITLE
Does not support opening notes when using vscode extension

### ISSUE THREAD

**Issue by 3v1n0** [MrCarroll/joplin-desktop-snap#16]
Trying to use this snap together to the [joplin vscode plugin](https://github.com/rxliuli/joplin-vscode-plugin) doesn't work properly:

- Notes are properly listed and can be managed (removed, renamed, created)
- Opening a note fails.

The only log I get is in vscode:

```
[2021-04-19 14:38:49.930] [renderer3] [error] Running the contributed command: 'joplinNote.openNote' failed.: Error: Running the contributed command: 'joplinNote.openNote' failed.
	at S._executeContributedCommand (/snap/code/62/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:86:106930)
	at runMicrotasks (<anonymous>)
	at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

But since using the upstream AppImage it works, I assume there's some problem with the snap. There's no audit log in `dmesg` either.

**Comment by rxliuli** [comment]
I guess it is related to another question: https://github.com/MrCarroll/joplin-desktop-snap/issues/9

**Comment by JGCarroll** [comment]
I've just tried it out now and it works for me, I'd double check to ensure the web clipper port and authentication tokens are correct, and that the `Program Profile Path` value is correct, it should be `/home/user/snap/joplin-desktop/current/.config/joplin-desktop`.

What I am noticing is that VSCode opens what I assume is a xdg-desktop-portals interface to select the editor. What's annoying with these portals is that after selecting the same editor 3 times, it remembers the result and stops asking, usually this would be a good thing to not nag the user, but I'm wondering if it's possible that your system is trying to open a markdown editor that doesn't exist rather than actually open in VSCode. Potentially then it's related to this https://github.com/flatpak/xdg-desktop-portal/issues/472

I don't imagine #9 is blocking the vscode plugin since VSCode itself is unconfined and all the logic happens over the network with the webclipper service, which the sandboxing doesn't interfere with.

**Comment by 3v1n0** [comment]
> I've just tried it out now and it works for me, I'd double check to ensure the web clipper port and authentication tokens are correct, and that the `Program Profile Path` value is correct, it should be `/home/user/snap/joplin-desktop/current/.config/joplin-desktop`.

Yeah, both were correct.

I'm not sure about the portal thingy, as when i tried running the upstream AppImage with `HOME=$HOME/snap/joplin-desktop/current ./Joplin-1.7.11.AppImage` it worked, so I assume in this case the vscode's snap should still require the user for input, or not?

**Comment by 3v1n0** [comment]
@rxliuli maybe is there a way I can debug this better from the vscode (extension) side?

**Comment by JGCarroll** [comment]
So the behaviour I'm seeing is really weird,

With either the snap or .deb installation of VSCode, the vscode-joplin plugin will open Gedit if using the AppImage, or use xdg-portals and hence present a choice if using the Joplin Snap. I can't actually get it to open Joplins notes in vscode with the AppImage version at all.

This is observed on a clean VM of Ubuntu 20.04, with nothing configured apart from VSC and Joplin.

So the answer appears to be no, it isn't assured that you're using the portals just because the AppImage works. You'd need to make sure the xdg-desktop-portals package for your system is actually installed which is default for a lot of systems (Ubuntu/Fedora/Debian/etc) but might not be as you go into more nicher territories, say Arch/I3 etc.

But I absolutely can't explain why Code is using xdg-portals at all since it's unconfined and doesn't have to, or why it seems to only do it when Joplin is running as a snap. It's very unusual behaviour.

**Comment by rxliuli** [comment]
Maybe you can try the following command to open the notes in VSCode. If so, it is likely to be a problem with the plug-in itself. @3v1n0 

```sh
curl --location --request POST 'http://localhost:27584/services/externalEditWatcher?token=***' \
--header 'Content-Type: application/json' \
--data-raw '{
    "action": "openAndWatch",
    "noteId": "257f6a9dacc1409580ee526d50ac4d49"
}'
```

ref: https://github.com/rxliuli/joplin-vscode-plugin/issues/37#issuecomment-820047247

**Comment by 3v1n0** [comment]
Ok...

So, using this with my note I was getting:

```
{"error":"Internal Server Error: spawn /snap/bin/code ENOENT\n\nCommand was: \"/snap/bin/code\" /home/marco/snap/joplin-desktop/7/.config/joplin-desktop/edit-16ede72c66bf4d629ab2be5277b90cb1.md: \n\nError: spawn /snap/bin/code ENOENT\n\nCommand was: \"/snap/bin/code\" /home/marco/snap/joplin-desktop/7/.config/joplin-desktop/edit-16ede72c66bf4d629ab2be5277b90cb1.md\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:267:19)\n    at onErrorNT (internal/child_process.js:469:16)\n    at processTicksAndRejections (internal/process/task_queues.js:84:21)"}%
```

This is because I was setting into the Joplin desktop to use `/snap/bin/code` as my editor, now setting it to default, I get the file opened with another editor... That isn't code.

So the issue is indeed related to https://github.com/MrCarroll/joplin-desktop-snap/issues/16#issuecomment-822517409

At the same time the extension shouldn't rely on the default editor configured in the Joplin desktop, but I guess for this it would require some other API...

**Comment by rxliuli** [comment]
> At the same time the extension shouldn't rely on the default editor configured in the Joplin desktop, but I guess for this it would require some other API...

Yes, it is currently completely dependent on the behavior of joplin action api, action api will use the default editor configured by joplin-desktop.

> Maybe at some point in the future I have to implement these logics manually: https://github.com/rxliuli/joplin-vscode-plugin/issues/39#issuecomment-822423575

**Comment by JGCarroll** [comment]
Assuming setting it to default empty value works (equivilent to setting it to `xdg-open`), if it's not asking you what to open but instead just opens something else, it's because the desktop portals have already associated a default choice for .md files since you've previously clicked on the same editor 3 times for .md files.  There's no other mechanism a strict snap would have to trigger another editor to open.

I'm trying to find a way to manually reset the assocations but I'm not having much success.

**Comment by JGCarroll** [comment]
Try removing/renaming `$HOME/.local/share/flatpak/db/desktop-used-apps`, then log in and out of the desktop session (or the old values are cached). This seemingly resets the associations and in theory the snap (and VSCode) should start asking you for a choice to open .md files with again, after using VSCode 3 times successively it should skip asking and work as expected, opening the files from the Joplin plugin in VSC will actually open them in VSC, and pressing control-e in Joplin itself will open VSCode too

**Comment by 3v1n0** [comment]
So, based on this I tried something else and it worked equally, without having to reset the db:
 - In the host I've changed the default for .md files to vscode
 - Once that was done, opening from Joplin was asking for the preferred application
   -  As you suggested, after the 3rd time it just used the default.

Changing again the default application for .md files to something else (i.e. gedit) made the prompt to appear again.

So, this is indeed a workaroud. Not sure we should close the bug though, as having an editor depending on the context may be preferred (i.e. it's fine to open code in this scenarios, but otherwise I'd prefer to use gedit).

**Comment by JGCarroll** [comment]
I don't mind leaving bugs open if they're valid faults, I don't get enough activity to worry about the issue tracker so if it benefits people by being visible all the better.

In terms of what's actionable though, the available API's give no influence to what takes responsibility for opening a file/protocol, it'd need to be tackled with the portals themseles. There's also the Flatpak portal which allows a confined application to call an unconfined application explicitly, but Snapcraft doesn't seem to want to adopt it (it's absolutely trivial to violate all the sandboxing if it's usable). Disregarding security, this does give the Flatpak a clear advantage for people who want absolute control over the external editor experience.

On the plus side if I remember correctly, the stored values are remembered per application, so setting .md to open in VSCode will only apply to this snap specifically and not automatically apply to any other applications that might also happen to request opening .md files.

**Comment by 3v1n0** [comment]
> On the plus side if I remember correctly, the stored values are remembered per application, so setting .md to open in VSCode will only apply to this snap specifically and not automatically apply to any other applications that might also happen to request opening .md files.

Sure, that's indeed a better thing, won't solve the case where one may want to use another editor for casual editing while vscode when called from the extension.

But indeed this is something that isn't related to the snap itself but more to the way the vscode extension acts and to the lack of APIs on the clipper service to do this

**Comment by JGCarroll** [comment]
It looks like this should be fixed with version 0.7 of the plugin, based on the changelog here. The snap would still need to be updated to Joplin version 2.8, but if it works with the flatpak, I'd expect it to work the same here.

https://discourse.joplinapp.org/t/joplin-vscode-plugin-roadmap/11335/29

**Comment by rxliuli** [comment]
> It looks like this should be fixed with version 0.7 of the plugin, based on the changelog here. The snap would still need to be updated to Joplin version 2.8, but if it works with the flatpak, I'd expect it to work the same here.
> 
> https://discourse.joplinapp.org/t/joplin-vscode-plugin-roadmap/11335/29

Essentially, version 0.7 only communicates with joplin through http, and no longer relies on joplin itself to open external editors, so it should not be restricted by the sandbox

**Comment by JGCarroll** [comment]
I've published the 2.8 release and so when users receive the update for both the snap and plugin, this should be solved.
